src/Npgsql/NpgsqlReadBuffer.cs

```diff
@@ -185,36 +188,64 @@ async Task EnsureLong()
-                    // _timeoutCts
-                    _timeoutCts.Stop();
-                    switch (e)
+                    catch (Exception e)
                     {
-                    // User requested the cancellation
-                    case OperationCanceledException _ when (cancellationToken.IsCancellationRequested):
-                        throw dontBreakOnCancellation ? e : Connector.Break(e);
-                    // Read timeout
-                    case OperationCanceledException _:
-                    // Note that mono throws SocketException with the wrong error (see #1330)
-                    case IOException _ when (e.InnerException as SocketException)?.SocketErrorCode ==
-                                            (Type.GetType("Mono.Runtime") == null ? SocketError.TimedOut : SocketError.WouldBlock):
-                        Debug.Assert(e is OperationCanceledException ? async : !async);
-                        throw new NpgsqlException("Exception while reading from stream", new TimeoutException("Timeout during reading attempt"));
+                        // Stopping twice (in case the previous Stop() call succeeded) doesn't hurt.
+                        // Not stopping will cause an assertion failure in debug mode when we call Start() the next time.
+                        // We can't stop in a finally block because Connector.Break() will dispose the buffer and the contained
+                        // _timeoutCts
+                        Cts.Stop();
+                        switch (e)
+                        {
+                        // User requested the cancellation
+                        case OperationCanceledException _ when (cancellationToken.IsCancellationRequested):
+                            throw readingNotifications ? e : Connector.Break(e);
+                        // Read timeout
+                        case OperationCanceledException _:
+                        // Note that mono throws SocketException with the wrong error (see #1330)
+                        case IOException _ when (e.InnerException as SocketException)?.SocketErrorCode ==
+                                                (Type.GetType("Mono.Runtime") == null ? SocketError.TimedOut : SocketError.WouldBlock):
+                            Debug.Assert(e is OperationCanceledException ? async : !async);
+
+                            if (!isCancellation && !readingNotifications)
+                            {
+                                var skipCancellationWait = false;
+
+                                try
+                                {
+                                    Connector.CancelRequest(throwExceptions: true, requestedByUser: false);
+                                }
+                                catch
+                                {
+                                    skipCancellationWait = true;
+                                }
+
+                                if (!skipCancellationWait)
```

Comment:
Can't this block just go into try try after Connector.CancelRequest, this way it gets naturally skipped if there's an exception?